diff --git a/include/PR/gbi.h b/include/PR/gbi.h
index ebfab07..c2db7c8 100644
--- a/include/PR/gbi.h
+++ b/include/PR/gbi.h
@@ -22,6 +22,20 @@
 
 #include <PR/ultratypes.h>
 
+/*
+ * Converts texel coordinates to S and T values
+ */
+// Bilerp/Average filter
+#define TC_B_OFF	-0.5
+#define TC_B(p)		(((p) + TC_B_OFF) * 32)
+// #define TC_B(p)		((((p) > TC_B_OFF) ? ((p) - TC_B_OFF) : (p)) * 32)
+#define ST_B(s, t)	{TC_B(s), TC_B(t)}
+
+// Point filter
+#define TX_P_OFF	0.25
+#define TC_P(p)		(((p) + TX_P_OFF) * 32)
+#define ST_P(s, t)	{TC_P(s), TC_P(t)}
+
 /*
  * To use the F3DEX ucodes, define F3DEX_GBI before include this file.
  *
@@ -687,10 +701,21 @@
 #define	G_ZS_PRIM		(1 << G_MDSFT_ZSRCSEL)
 
 /* G_SETOTHERMODE_L gSetRenderMode */
+#ifdef DISABLE_AA
+#define	AA_EN		0x0
+#else
 #define	AA_EN		0x8
+#endif
+
 #define	Z_CMP		0x10
 #define	Z_UPD		0x20
+
+#ifdef DISABLE_AA
+#define	IM_RD		0x00
+#else
 #define	IM_RD		0x40
+#endif
+
 #define	CLR_ON_CVG	0x80
 #define	CVG_DST_CLAMP	0
 #define	CVG_DST_WRAP	0x100
@@ -717,6 +742,14 @@
 #define	G_BL_1		2
 #define	G_BL_0		3
 
+#ifdef DISABLE_AA
+#define AA_DEF
+#define RD_DEF
+#else
+#define AA_DEF AA_EN |
+#define RD_DEF IM_RD |
+#endif
+
 #define	GBL_c1(m1a, m1b, m2a, m2b)	\
 	(m1a) << 30 | (m1b) << 26 | (m2a) << 22 | (m2b) << 18
 #define	GBL_c2(m1a, m1b, m2a, m2b)	\
@@ -943,9 +976,72 @@
 	RM_AA_ZB_XLU_SURF(clk) | Z_UPD
 
 
-#define	G_RM_AA_ZB_OPA_SURF	RM_AA_ZB_OPA_SURF(1)
+#ifdef DISABLE_AA
+
+#define	G_RM_AA_ZB_OPA_SURF		RM_ZB_OPA_SURF(1)
+#define	G_RM_AA_ZB_OPA_SURF2	RM_ZB_OPA_SURF(2)
+#define	G_RM_AA_ZB_XLU_SURF		RM_ZB_XLU_SURF(1)
+#define	G_RM_AA_ZB_XLU_SURF2	RM_ZB_XLU_SURF(2)
+#define	G_RM_AA_ZB_OPA_DECAL	RM_ZB_OPA_DECAL(1)
+#define	G_RM_AA_ZB_OPA_DECAL2	RM_ZB_OPA_DECAL(2)
+#define	G_RM_AA_ZB_XLU_DECAL	RM_ZB_XLU_DECAL(1)
+#define	G_RM_AA_ZB_XLU_DECAL2	RM_ZB_XLU_DECAL(2)
+#define	G_RM_AA_ZB_OPA_INTER	RM_AA_ZB_OPA_INTER(1)
+#define	G_RM_AA_ZB_OPA_INTER2	RM_AA_ZB_OPA_INTER(2)
+#define	G_RM_AA_ZB_XLU_INTER	RM_AA_ZB_XLU_INTER(1)
+#define	G_RM_AA_ZB_XLU_INTER2	RM_AA_ZB_XLU_INTER(2)
+#define	G_RM_AA_ZB_XLU_LINE		RM_AA_ZB_XLU_LINE(1)
+#define	G_RM_AA_ZB_XLU_LINE2	RM_AA_ZB_XLU_LINE(2)
+#define	G_RM_AA_ZB_DEC_LINE		RM_AA_ZB_DEC_LINE(1)
+#define	G_RM_AA_ZB_DEC_LINE2	RM_AA_ZB_DEC_LINE(2)
+#define	G_RM_AA_ZB_TEX_EDGE		RM_AA_ZB_TEX_EDGE(1)
+#define	G_RM_AA_ZB_TEX_EDGE2	RM_AA_ZB_TEX_EDGE(2)
+#define	G_RM_AA_ZB_TEX_INTER	RM_AA_ZB_TEX_INTER(1)
+#define	G_RM_AA_ZB_TEX_INTER2	RM_AA_ZB_TEX_INTER(2)
+#define	G_RM_AA_ZB_SUB_SURF		RM_AA_ZB_SUB_SURF(1)
+#define	G_RM_AA_ZB_SUB_SURF2	RM_AA_ZB_SUB_SURF(2)
+#define	G_RM_AA_ZB_PCL_SURF		RM_ZB_PCL_SURF(1)
+#define	G_RM_AA_ZB_PCL_SURF2	RM_ZB_PCL_SURF(2)
+#define	G_RM_AA_ZB_OPA_TERR		RM_AA_ZB_OPA_TERR(1)
+#define	G_RM_AA_ZB_OPA_TERR2	RM_AA_ZB_OPA_TERR(2)
+#define	G_RM_AA_ZB_TEX_TERR		RM_AA_ZB_TEX_TERR(1)
+#define	G_RM_AA_ZB_TEX_TERR2	RM_AA_ZB_TEX_TERR(2)
+#define	G_RM_AA_ZB_SUB_TERR		RM_AA_ZB_SUB_TERR(1)
+#define	G_RM_AA_ZB_SUB_TERR2	RM_AA_ZB_SUB_TERR(2)
+
+#define	G_RM_RA_ZB_OPA_SURF		RM_RA_ZB_OPA_SURF(1)
+#define	G_RM_RA_ZB_OPA_SURF2	RM_RA_ZB_OPA_SURF(2)
+#define	G_RM_RA_ZB_OPA_DECAL	RM_RA_ZB_OPA_DECAL(1)
+#define	G_RM_RA_ZB_OPA_DECAL2	RM_RA_ZB_OPA_DECAL(2)
+#define	G_RM_RA_ZB_OPA_INTER	RM_RA_ZB_OPA_INTER(1)
+#define	G_RM_RA_ZB_OPA_INTER2	RM_RA_ZB_OPA_INTER(2)
+
+#define	G_RM_AA_OPA_SURF		RM_OPA_SURF(1)
+#define	G_RM_AA_OPA_SURF2		RM_OPA_SURF(2)
+#define	G_RM_AA_XLU_SURF		RM_XLU_SURF(1)
+#define	G_RM_AA_XLU_SURF2		RM_XLU_SURF(2)
+#define	G_RM_AA_XLU_LINE		RM_AA_XLU_LINE(1)
+#define	G_RM_AA_XLU_LINE2		RM_AA_XLU_LINE(2)
+#define	G_RM_AA_DEC_LINE		RM_AA_DEC_LINE(1)
+#define	G_RM_AA_DEC_LINE2		RM_AA_DEC_LINE(2)
+#define	G_RM_AA_TEX_EDGE		RM_TEX_EDGE(1)
+#define	G_RM_AA_TEX_EDGE2		RM_TEX_EDGE(2)
+#define	G_RM_AA_SUB_SURF		RM_AA_SUB_SURF(1)
+#define	G_RM_AA_SUB_SURF2		RM_AA_SUB_SURF(2)
+#define	G_RM_AA_PCL_SURF		RM_PCL_SURF(1)
+#define	G_RM_AA_PCL_SURF2		RM_PCL_SURF(2)
+#define	G_RM_AA_OPA_TERR		RM_AA_OPA_TERR(1)
+#define	G_RM_AA_OPA_TERR2		RM_AA_OPA_TERR(2)
+#define	G_RM_AA_TEX_TERR		RM_AA_TEX_TERR(1)
+#define	G_RM_AA_TEX_TERR2		RM_AA_TEX_TERR(2)
+#define	G_RM_AA_SUB_TERR		RM_AA_SUB_TERR(1)
+#define	G_RM_AA_SUB_TERR2		RM_AA_SUB_TERR(2)
+
+#else
+
+#define	G_RM_AA_ZB_OPA_SURF		RM_AA_ZB_OPA_SURF(1)
 #define	G_RM_AA_ZB_OPA_SURF2	RM_AA_ZB_OPA_SURF(2)
-#define	G_RM_AA_ZB_XLU_SURF	RM_AA_ZB_XLU_SURF(1)
+#define	G_RM_AA_ZB_XLU_SURF		RM_AA_ZB_XLU_SURF(1)
 #define	G_RM_AA_ZB_XLU_SURF2	RM_AA_ZB_XLU_SURF(2)
 #define	G_RM_AA_ZB_OPA_DECAL	RM_AA_ZB_OPA_DECAL(1)
 #define	G_RM_AA_ZB_OPA_DECAL2	RM_AA_ZB_OPA_DECAL(2)
@@ -955,26 +1051,26 @@
 #define	G_RM_AA_ZB_OPA_INTER2	RM_AA_ZB_OPA_INTER(2)
 #define	G_RM_AA_ZB_XLU_INTER	RM_AA_ZB_XLU_INTER(1)
 #define	G_RM_AA_ZB_XLU_INTER2	RM_AA_ZB_XLU_INTER(2)
-#define	G_RM_AA_ZB_XLU_LINE	RM_AA_ZB_XLU_LINE(1)
+#define	G_RM_AA_ZB_XLU_LINE		RM_AA_ZB_XLU_LINE(1)
 #define	G_RM_AA_ZB_XLU_LINE2	RM_AA_ZB_XLU_LINE(2)
-#define	G_RM_AA_ZB_DEC_LINE	RM_AA_ZB_DEC_LINE(1)
+#define	G_RM_AA_ZB_DEC_LINE		RM_AA_ZB_DEC_LINE(1)
 #define	G_RM_AA_ZB_DEC_LINE2	RM_AA_ZB_DEC_LINE(2)
-#define	G_RM_AA_ZB_TEX_EDGE	RM_AA_ZB_TEX_EDGE(1)
+#define	G_RM_AA_ZB_TEX_EDGE		RM_AA_ZB_TEX_EDGE(1)
 #define	G_RM_AA_ZB_TEX_EDGE2	RM_AA_ZB_TEX_EDGE(2)
 #define	G_RM_AA_ZB_TEX_INTER	RM_AA_ZB_TEX_INTER(1)
 #define	G_RM_AA_ZB_TEX_INTER2	RM_AA_ZB_TEX_INTER(2)
-#define	G_RM_AA_ZB_SUB_SURF	RM_AA_ZB_SUB_SURF(1)
+#define	G_RM_AA_ZB_SUB_SURF		RM_AA_ZB_SUB_SURF(1)
 #define	G_RM_AA_ZB_SUB_SURF2	RM_AA_ZB_SUB_SURF(2)
-#define	G_RM_AA_ZB_PCL_SURF	RM_AA_ZB_PCL_SURF(1)
+#define	G_RM_AA_ZB_PCL_SURF		RM_AA_ZB_PCL_SURF(1)
 #define	G_RM_AA_ZB_PCL_SURF2	RM_AA_ZB_PCL_SURF(2)
-#define	G_RM_AA_ZB_OPA_TERR	RM_AA_ZB_OPA_TERR(1)
+#define	G_RM_AA_ZB_OPA_TERR		RM_AA_ZB_OPA_TERR(1)
 #define	G_RM_AA_ZB_OPA_TERR2	RM_AA_ZB_OPA_TERR(2)
-#define	G_RM_AA_ZB_TEX_TERR	RM_AA_ZB_TEX_TERR(1)
+#define	G_RM_AA_ZB_TEX_TERR		RM_AA_ZB_TEX_TERR(1)
 #define	G_RM_AA_ZB_TEX_TERR2	RM_AA_ZB_TEX_TERR(2)
-#define	G_RM_AA_ZB_SUB_TERR	RM_AA_ZB_SUB_TERR(1)
+#define	G_RM_AA_ZB_SUB_TERR		RM_AA_ZB_SUB_TERR(1)
 #define	G_RM_AA_ZB_SUB_TERR2	RM_AA_ZB_SUB_TERR(2)
 
-#define	G_RM_RA_ZB_OPA_SURF	RM_RA_ZB_OPA_SURF(1)
+#define	G_RM_RA_ZB_OPA_SURF		RM_RA_ZB_OPA_SURF(1)
 #define	G_RM_RA_ZB_OPA_SURF2	RM_RA_ZB_OPA_SURF(2)
 #define	G_RM_RA_ZB_OPA_DECAL	RM_RA_ZB_OPA_DECAL(1)
 #define	G_RM_RA_ZB_OPA_DECAL2	RM_RA_ZB_OPA_DECAL(2)
@@ -1002,6 +1098,8 @@
 #define	G_RM_AA_SUB_TERR	RM_AA_SUB_TERR(1)
 #define	G_RM_AA_SUB_TERR2	RM_AA_SUB_TERR(2)
 
+#endif
+
 #define	G_RM_RA_OPA_SURF	RM_RA_OPA_SURF(1)
 #define	G_RM_RA_OPA_SURF2	RM_RA_OPA_SURF(2)
 
diff --git a/include/PatchConfig/config_Graphics.h b/include/PatchConfig/config_Graphics.h
new file mode 100644
index 0000000..94a9743
--- /dev/null
+++ b/include/PatchConfig/config_Graphics.h
@@ -0,0 +1,6 @@
+ /********************
+ * GRAPHICS SETTINGS *
+ ********************/
+
+// Toggles if Anti Aliasing is enabled or disabled (Improves console performance but causes visible seams between unconnected geometry).
+// #define DISABLE_AA
diff --git a/include/config.h b/include/config.h
index 044b069..09f5639 100644
--- a/include/config.h
+++ b/include/config.h
@@ -50,4 +50,8 @@
 #define BORDER_HEIGHT 0
 #endif
 
+//My Patches are now inside a Folder called PatchConfig so they dont cause issues with the
+//Vanilla Decomp Config File and also to make it look more clean
+#include "PatchConfig/config_Graphics.h"
+
 #endif // CONFIG_H
diff --git a/include/macros.inc b/include/macros.inc
index 3cb002a..838c32a 100644
--- a/include/macros.inc
+++ b/include/macros.inc
@@ -1,4 +1,4 @@
-// Assembly Macros
+/* Assembly Macros */
 
 .set K0BASE, 0x80000000
 .set K1BASE, 0xA0000000
diff --git a/include/sm64.h b/include/sm64.h
index 88446c9..3ca72ed 100644
--- a/include/sm64.h
+++ b/include/sm64.h
@@ -49,6 +49,8 @@
 #define LAYER_TRANSPARENT_DECAL 6
 #define LAYER_TRANSPARENT_INTER 7
 
+#define LAYER_FIRST                         LAYER_FORCE
+
 #define INPUT_NONZERO_ANALOG         0x0001
 #define INPUT_A_PRESSED              0x0002
 #define INPUT_OFF_FLOOR              0x0004
